packages/core/src/modules/drpc/DRPCMessagesApi.ts

```diff
@@ -0,0 +1,173 @@
+    messageSender: MessageSender,
+    connectionService: ConnectionService,
+    agentContext: AgentContext
+  ) {
+    this.drpcMessageService = drpcMessageService
+    this.messageSender = messageSender
+    this.connectionService = connectionService
+    this.agentContext = agentContext
+    this.registerMessageHandlers(messageHandlerRegistry)
+
+    // listen for incoming drpc requests and forward them to appropriate handlers
+    this.drpcMessageService.createMessageListener(async ({ message, drpcMessageRecord }) => {
+      if (message instanceof DRPCRequestMessage && drpcMessageRecord.role === DRPCMessageRole.Receiver) {
+        const requests = Array.isArray(message.request) ? message.request : [message.request]
+        const futures = []
+        for (const request of requests) {
+          futures.push(this.handleRequest(request))
+        }
+        const responses = await Promise.all(futures)
+        try {
+          await this.sendDRPCResponse(
+            drpcMessageRecord.connectionId,
+            message.id,
+            responses.length === 1 ? responses[0] : responses
+          )
+        } catch {
+          await this.sendDRPCResponse(drpcMessageRecord.connectionId, message.id, {
+            jsonrpc: '2.0',
+            id: null,
+            error: { code: DRPCErrorCode.INTERNAL_ERROR, message: 'Internal error', data: 'Error sending response' },
+          })
+        }
+      }
+    })
+  }
+
+  private async handleRequest(message: DRPCRequestObject) {
+    if (!isValidDRPCRequestObject(message)) {
+      return {
+        jsonrpc: '2.0',
+        id: message.id,
+        error: { code: DRPCErrorCode.INVALID_REQUEST, message: 'Invalid request' },
+      }
+    }
+    const handler = this.drpcMethodHandlers.get(message.method)
+    if (!handler) {
+      return {
+        jsonrpc: '2.0',
+        id: message.id,
+        error: { code: DRPCErrorCode.METHOD_NOT_FOUND, message: 'Method not found' },
+      }
+    }
+
+    try {
+      let response: DRPCResponseObject | Record<string, never> = await handler(message)
+      if (message.id === null) {
+        response = {}
+      }
+      return response
+    } catch (error) {
+      return {
+        jsonrpc: '2.0',
+        id: message.id,
+        error: { code: DRPCErrorCode.INTERNAL_ERROR, message: 'Internal error', data: error.message },
+      }
+    }
+  }
+
+  public async sendDRPCRequest(connectionId: string, request: DRPCRequest): Promise<DRPCResponse> {
+    const messageId = uuid()
+    const connection = await this.connectionService.getById(this.agentContext, connectionId)
+    try {
+      const { message: drpcMessage, record: drpcMessageRecord } = await this.drpcMessageService.createRequestMessage(
+        this.agentContext,
+        request,
+        connection,
+        messageId
+      )
+      await this.sendDRPCMessage(connection, drpcMessage, drpcMessageRecord)
+    } catch {
+      throw new Error('Invalid DRPC Request')
```

Comment:
Maybe we can leave sendDRPCMessage throw its error? Because it can be of different natures (e.g. network error).